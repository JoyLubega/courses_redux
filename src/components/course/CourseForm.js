import React from 'react';
import PropTypes from 'prop-types'
import TextInput from '../common/TextInput';
import SelectInput from '../common/SelectInput';

const CourseForm = (props) => {
    const { course, allAuthors, onSave, onChange, saving, errors, onDelete}= props;
    let urlParams = new URLSearchParams(window.location.search);
    let deleted = urlParams.get('delete');
  return (
    <form className="container">
        {deleted ? <h1>Add Course</h1> :
        <h1>Edit Course</h1>
        }
      
      <TextInput
        name="title"
        label="Title"
        value={course.title}
        onChange={onChange}
        error={errors.title}/>

      <SelectInput
        name="authorId"
        label="Author"
        value={course.authorId}
        defaultOption="Select Author"
        options={allAuthors}
        onChange={onChange} 
        error={errors.authorId}/>

      <TextInput
        name="category"
        label="Category"
        value={course.category}
        onChange={onChange}
        error={errors.category}/>

      <TextInput
        name="length"
        label="Length"
        value={course.length}
        onChange={onChange}
        error={errors.length}/>

      <input
        type="submit"
        disabled={saving}
        value={saving ? 'Saving...' : 'Save'}
        className="btn btn-success"
        onClick={onSave}/>
        {}
        
        {deleted ? <div/>:
        <input
        type="submit"
        disabled={saving}
        value="Delete"
        className="btn btn-danger "
        onClick={onDelete}
        />
        }
    </form>
  );
};

CourseForm.propTypes = {
  course: PropTypes.object.isRequired,
  allAuthors: PropTypes.array.isRequired,
  onSave: PropTypes.func.isRequired,
  onDelete: PropTypes.func.isRequired,
  onChange: PropTypes.func.isRequired,
  saving: PropTypes.bool,
  errors: PropTypes.object
};

export default CourseForm;