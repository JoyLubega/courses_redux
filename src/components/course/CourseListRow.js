import React from 'react';
import PropTypes from 'prop-types'
import {Link} from 'react-router-dom';

const CourseListRow = (props) => {
  const { course} = props;
  return (
    <tr>
      <td><a href={course.watchHref} target="_blank">Watch</a></td>
      <td><a href={'/course/' + course.id}>{course.title}</a></td>
      <td>{course.authorId}</td>
      <td>{course.category}</td>
      <td>{course.length}</td>
    </tr>
  );
};
CourseListRow.propTypes = {
  course: PropTypes.object.isRequired
};
export default CourseListRow;