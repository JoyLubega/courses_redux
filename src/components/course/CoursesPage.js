import React from 'react';
import PropTypes from 'prop-types'
import { bindActionCreators } from 'redux';
import {connect} from 'react-redux';
import * as courseActions from '../../actions/courseActions';
import CourseList from './CourseList';


 export class CoursesPage extends React.Component {
  constructor(props, context) {
    super(props, context);
  this.redirectToAddCoursePage = this.redirectToAddCoursePage.bind(this)
  }
  courseRow(course, index){
    return <div key={index}>{course.title}</div>;
  }
 
  redirectToAddCoursePage(event,id){ 
    this.props.history.push("/course?delete=false");
  } 
  render() {
    const {courses} = this.props;
    return (
      <div className="container">
        <h1>Courses</h1>  
        <button type="button"
               className="btn btn-warning"
               onClick={this.redirectToAddCoursePage}
              >Add Course</button>
              <br/>
        <CourseList courses={courses}/>
      </div>
    );
  }
}
CoursesPage.propTypes = {
  actions: PropTypes.object.isRequired,
  courses: PropTypes.array.isRequired
};
const mapStateToProps=(state, ownProps)=> {
  return {
    courses: state.courses
  };
};
const mapDispatchToProps=(dispatch)=> {
  return {
    actions: bindActionCreators(courseActions, dispatch)
  };
};
export default connect(mapStateToProps, mapDispatchToProps)(CoursesPage);