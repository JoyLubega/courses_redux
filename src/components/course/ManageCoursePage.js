import React from 'react';
import PropTypes from 'prop-types'
import {connect} from 'react-redux';
import {bindActionCreators} from 'redux';
import * as courseActions from '../../actions/courseActions';
import CourseForm from './CourseForm';
import Notifications, {notify} from 'react-notify-toast';

let danger = { background: 'Red', text: "White" };
let success = { background: 'Green', text: "White" };
class ManageCoursePage extends React.Component {
  constructor(props, context) {
    super(props, context);
    this.state = {
      course: Object.assign({}, this.props.course),
      errors: {},
      saving: false
    };
    this.saveOrUpdateCourse = this.saveOrUpdateCourse.bind(this);
    this.deleteCourse= this.deleteCourse.bind(this);
    this.updateCourseState = this.updateCourseState.bind(this);
  }
  componentWillReceiveProps(nextProps) {
    if (this.props.course.id !== nextProps.course.id) {
      this.setState({course: Object.assign({}, nextProps.course)});
    }
  }
  updateCourseState(event) {
    const field = event.target.name;
    let course = Object.assign({}, this.state.course);
    course[field] = event.target.value;
    return this.setState({course: course});
  }
  deleteCourse(event){
    event.preventDefault();
    this.props.actions.deleteCourse(this.state.course.id)
    .then(() => this.redirect())
    .catch(error => {
      notify.show(error, "custom", 5000, danger);
      });
  }
  saveOrUpdateCourse(event) {
    event.preventDefault();
    this.setState({ saving:true });
    this.props.actions.saveCourse(this.state.course)
        .then(() => this.redirect())
        .catch(error => {
          notify.show(error, "custom", 5000, danger);
            this.setState({saving: false});
          });       
}
redirect(){
    this.props.history.push('/courses'); 
}
  render() {
    
    return (
      <div>
      <Notifications />
      <CourseForm
        course={this.state.course}
        onChange={this.updateCourseState}
        onSave={this.saveOrUpdateCourse}
        errors={this.state.errors}
        allAuthors={this.props.authors}
        onDelete={this.deleteCourse}
        saving={this.state.saving}
      />
      </div>
    );
  }
}

ManageCoursePage.propTypes = {
  course: PropTypes.object.isRequired,
  authors: PropTypes.array.isRequired,
  actions: PropTypes.object.isRequired

};

ManageCoursePage.contextTypes = {
  router: PropTypes.object
};

const getCourseById = (courses, id) => {
    const course = courses.filter(course => course.id === id);
    if (course) return course[0]; //since filter returns an array, have to grab the first.
    return null;
  };

const mapStateToProps=(state, props)=> {
  
    const courseId = props.match.params.id; // this id is coming from  the path `/course/:id`
    let course = {id: '', watchHref: '', title: '', authorId: '', length: '', category: ''};
    if (courseId && state.courses.length > 0){
      
        course = getCourseById(state.courses, courseId);

    }    
    const authorsFormattedForDropdown = state.authors.map(author => {
        return {
            value: author.id,
            text: author.firstName +' '+ author.lastName
        };
    });
    
    return {
      course: course,
      authors: authorsFormattedForDropdown
  };
};

const mapDispatchToProps=(dispatch)=> {
  return {
    actions: bindActionCreators(courseActions, dispatch)
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(ManageCoursePage);