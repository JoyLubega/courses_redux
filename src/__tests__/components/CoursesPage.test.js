import React from 'react';
import { shallow} from 'enzyme';
import {CoursesPage, mapDispatchToProps, mapStateToProps } from '../../components/course/CoursesPage';
// import { mapStateToProps } from '../../components/course/ManageCoursePage';

const setup=() =>{
    const props = {
      courses:[],
      actions:{}
    }
    return shallow(<CoursesPage {...props} />);
}
  it('renders the CourseList component', () => {
    const wrapper = setup(true);
    expect(wrapper.find('table')).toMatchSnapshot();
  });
  it('mapDispatchToProps', () => {
    const dispatch = jest.fn();
    mapDispatchToProps(dispatch).actions;
    expect(dispatch.mock.calls).toMatchSnapshot();
  });
  it('mapDispatchToProps', () => {
    const dispatch = jest.fn();
    mapDispatchToProps(dispatch).actions;
    expect(dispatch.mock.calls).toMatchSnapshot();
  });
  it ('It renders mapStateToprops',()=>{
    const state={
      courses:[]
    }
    expect(mapStateToProps(state)).toEqual(state)
  });

describe('<CourseList /> component contains child componenets', () => {
  it('renders the CourseList component', () => {
    const wrapper = setup(true);
    expect(wrapper.find('CourseListRow')).toMatchSnapshot();
  });
});
