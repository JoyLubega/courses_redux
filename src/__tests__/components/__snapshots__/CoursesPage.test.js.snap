// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CourseList /> component contains child componenets renders the CourseList component 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): ShallowWrapper {
    Symbol(enzyme.__root__): [Circular],
    Symbol(enzyme.__unrendered__): <CoursesPage
      actions={Object {}}
      courses={Array []}
    />,
    Symbol(enzyme.__renderer__): Object {
      "batchedUpdates": [Function],
      "getNode": [Function],
      "render": [Function],
      "simulateError": [Function],
      "simulateEvent": [Function],
      "unmount": [Function],
    },
    Symbol(enzyme.__node__): Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h1>
            Courses
          </h1>,
          <button
            className="btn btn-warning"
            onClick={[Function]}
            type="button"
          >
            Add Course
          </button>,
          <br />,
          <CourseList
            courses={Array []}
          />,
        ],
        "className": "container",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Courses",
          },
          "ref": null,
          "rendered": "Courses",
          "type": "h1",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Add Course",
            "className": "btn btn-warning",
            "onClick": [Function],
            "type": "button",
          },
          "ref": null,
          "rendered": "Add Course",
          "type": "button",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": "br",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "courses": Array [],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
    Symbol(enzyme.__nodes__): Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <h1>
              Courses
            </h1>,
            <button
              className="btn btn-warning"
              onClick={[Function]}
              type="button"
            >
              Add Course
            </button>,
            <br />,
            <CourseList
              courses={Array []}
            />,
          ],
          "className": "container",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Courses",
            },
            "ref": null,
            "rendered": "Courses",
            "type": "h1",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Add Course",
              "className": "btn btn-warning",
              "onClick": [Function],
              "type": "button",
            },
            "ref": null,
            "rendered": "Add Course",
            "type": "button",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": "br",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "courses": Array [],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
    ],
    Symbol(enzyme.__options__): Object {
      "adapter": ReactSixteenAdapter {
        "options": Object {
          "enableComponentDidUpdateOnSetState": true,
          "lifecycles": Object {
            "componentDidUpdate": Object {
              "onSetState": true,
            },
            "getDerivedStateFromProps": true,
            "getSnapshotBeforeUpdate": true,
            "setState": Object {
              "skipsComponentDidUpdateOnNullish": true,
            },
          },
        },
      },
    },
  },
  Symbol(enzyme.__unrendered__): null,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): undefined,
  Symbol(enzyme.__nodes__): Array [],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__rootNodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h1>
            Courses
          </h1>,
          <button
            className="btn btn-warning"
            onClick={[Function]}
            type="button"
          >
            Add Course
          </button>,
          <br />,
          <CourseList
            courses={Array []}
          />,
        ],
        "className": "container",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Courses",
          },
          "ref": null,
          "rendered": "Courses",
          "type": "h1",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Add Course",
            "className": "btn btn-warning",
            "onClick": [Function],
            "type": "button",
          },
          "ref": null,
          "rendered": "Add Course",
          "type": "button",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": "br",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "courses": Array [],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
}
`;

exports[`mapDispatchToProps 1`] = `Array []`;

exports[`mapDispatchToProps 2`] = `Array []`;

exports[`renders the CourseList component 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): ShallowWrapper {
    Symbol(enzyme.__root__): [Circular],
    Symbol(enzyme.__unrendered__): <CoursesPage
      actions={Object {}}
      courses={Array []}
    />,
    Symbol(enzyme.__renderer__): Object {
      "batchedUpdates": [Function],
      "getNode": [Function],
      "render": [Function],
      "simulateError": [Function],
      "simulateEvent": [Function],
      "unmount": [Function],
    },
    Symbol(enzyme.__node__): Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h1>
            Courses
          </h1>,
          <button
            className="btn btn-warning"
            onClick={[Function]}
            type="button"
          >
            Add Course
          </button>,
          <br />,
          <CourseList
            courses={Array []}
          />,
        ],
        "className": "container",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Courses",
          },
          "ref": null,
          "rendered": "Courses",
          "type": "h1",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Add Course",
            "className": "btn btn-warning",
            "onClick": [Function],
            "type": "button",
          },
          "ref": null,
          "rendered": "Add Course",
          "type": "button",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": "br",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "courses": Array [],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
    Symbol(enzyme.__nodes__): Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <h1>
              Courses
            </h1>,
            <button
              className="btn btn-warning"
              onClick={[Function]}
              type="button"
            >
              Add Course
            </button>,
            <br />,
            <CourseList
              courses={Array []}
            />,
          ],
          "className": "container",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Courses",
            },
            "ref": null,
            "rendered": "Courses",
            "type": "h1",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Add Course",
              "className": "btn btn-warning",
              "onClick": [Function],
              "type": "button",
            },
            "ref": null,
            "rendered": "Add Course",
            "type": "button",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": "br",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "courses": Array [],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
    ],
    Symbol(enzyme.__options__): Object {
      "adapter": ReactSixteenAdapter {
        "options": Object {
          "enableComponentDidUpdateOnSetState": true,
          "lifecycles": Object {
            "componentDidUpdate": Object {
              "onSetState": true,
            },
            "getDerivedStateFromProps": true,
            "getSnapshotBeforeUpdate": true,
            "setState": Object {
              "skipsComponentDidUpdateOnNullish": true,
            },
          },
        },
      },
    },
  },
  Symbol(enzyme.__unrendered__): null,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): undefined,
  Symbol(enzyme.__nodes__): Array [],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__rootNodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h1>
            Courses
          </h1>,
          <button
            className="btn btn-warning"
            onClick={[Function]}
            type="button"
          >
            Add Course
          </button>,
          <br />,
          <CourseList
            courses={Array []}
          />,
        ],
        "className": "container",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Courses",
          },
          "ref": null,
          "rendered": "Courses",
          "type": "h1",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Add Course",
            "className": "btn btn-warning",
            "onClick": [Function],
            "type": "button",
          },
          "ref": null,
          "rendered": "Add Course",
          "type": "button",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": "br",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "courses": Array [],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
}
`;
